src/Collapse/include/gudhi/Flag_complex_edge_collapser.h

```diff
@@ -15,366 +16,309 @@
-   */
-  template<typename FilteredEdgeRange>
-  Flag_complex_edge_collapser(const FilteredEdgeRange& edges)
-  : f_edge_vector_(std::begin(edges), std::end(edges)) { }
-
-  /** \brief Performs edge collapse in a increasing sequence of the filtration value.
-   *
-   * \tparam filtered_edge_output is a functor that is called on the output edges, in non-decreasing order of
-   * filtration, as filtered_edge_output(u, v, f) where u and v are Vertex_handle representing the extremities of the
-   * edge, and f is its new Filtration_value.
-   */
-  template<typename FilteredEdgeOutput>
-  void process_edges(FilteredEdgeOutput filtered_edge_output) {
-    // Sort edges
-    auto sort_by_filtration = [](const Filtered_edge& edge_a, const Filtered_edge& edge_b) -> bool
+  template<class FilteredEdgeRange, class Delay>
+  void process_edges(FilteredEdgeRange const& edges, Delay&& delay) {
     {
-      return (std::get<2>(edge_a) < std::get<2>(edge_b)); 
-    };
+      Vertex maxi = 0, maxj = 0;
+      for(auto& fe : edges) {
+        Vertex i = std::get<0>(fe);
+        Vertex j = std::get<1>(fe);
+        if (i > maxi) maxi = i;
+        if (j > maxj) maxj = j;
+      }
+      num_vertices = std::max(maxi, maxj) + 1;
+    }
 
-#ifdef GUDHI_USE_TBB
-    tbb::parallel_sort(f_edge_vector_.begin(), f_edge_vector_.end(), sort_by_filtration);
+    read_edges(edges);
+
+    boost::container::flat_set<Vertex> e_ngb;
+    e_ngb.reserve(num_vertices);
+    std::vector<std::pair<Filtration_value, Vertex>> e_ngb_later;
+    for(auto&e:edges) {
+      {
+        Vertex u = std::get<0>(e);
+        Vertex v = std::get<1>(e);
+        Filtration_value input_time = std::get<2>(e);
+        auto time = delay(input_time);
+        auto start_time = time;
+        e_ngb.clear();
+        e_ngb_later.clear();
+        common_neighbors(e_ngb, e_ngb_later, u, v, time);
+        // If we identify a good candidate (the first common neighbor) for being a dominator of e until infinity,
+        // we could check that a bit more cheaply. It does not seem to help though.
+        auto cmp1=[](auto const&a, auto const&b){return a.first > b.first;};
+        auto e_ngb_later_begin=e_ngb_later.begin();
+        auto e_ngb_later_end=e_ngb_later.end();
+        bool heapified = false;
+
+        bool dead = false;
+        while(true) {
+          Vertex dominator = -1;
+          // special case for size 1
+          // if(e_ngb.size()==1){dominator=*e_ngb.begin();}else
+          // It is tempting to test the dominators in increasing order of filtration value, which is likely to reduce
+          // the number of calls to is_dominated_by before finding a dominator, but sorting, even partially / lazily,
+          // is very expensive.
+          for(auto c : e_ngb){
+            if(is_dominated_by(e_ngb, c, time)){
+              dominator = c;
+              break;
+            }
+          }
+          if(dominator==-1) break;
+          bool still_dominated;
```

Comment:
Why would it need a default value? It is set in the loop before being used. And it is declared outside because it is used in the loop condition. Initializing it before the loop looks redundant, and could mislead a reader into thinking that this initial value has a meaning.

Reply:
I always find it better to initialize variables to avoid [risk of used before set](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-init)